File: LeetCode/java/131_Palindrome_Partitioning.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    List<List<String>> answer = new ArrayList<>();

    private boolean isPalindrome(String s) {
        StringBuilder rev = new StringBuilder(s);
        rev.reverse();

        if (rev.toString().equals(s)) {
            return true;
        }

        return false;
    }

    private void check(String s, int start, int[][] dp, List<String> list) {
        if (start == s.length()) {
            List<String> list2 = new ArrayList<>();

            for (String st : list) {
                list2.add(st);
            }

            answer.add(list2);
            return;
        }

        for (int end = start; end < s.length(); end++) {
            if (dp[start][end] == -1) {
                if (isPalindrome(s.substring(start, end + 1))) {
                    dp[start][end] = 1;
                } else {
                    dp[start][end] = 0;
                }
            }

            if (dp[start][end] == 1) {
                list.add(s.substring(start, end + 1));

                check(s, end + 1, dp, list);
                list.remove(list.size() - 1);
            }
        }
    }

    public List<List<String>> partition(String s) {
        int[][] dp = new int[s.length()][s.length()];
        List<String> list = new ArrayList<>();

        for (int i = 0; i < dp.length; i++) {
            Arrays.fill(dp[i], -1);
        }

        check(s, 0, dp, list);

        return answer;
    }
}

File: LeetCode/cpp/131_Palindrome_Partitioning.cpp
class Solution {
public:
	bool isPalindrome(string s) {
		string rev(s);
		reverse(rev.begin(), rev.end());

		if (s == rev) {
			return true;
		}

		return false;
	}

	void check(string s, int start, vector<vector<string>>& answer, vector<vector<int>>& dp, vector<string>& v) {
		if (start == s.length()) {
			answer.push_back(v);
			return;
		}

		for (int end = start; end < s.length(); end++) {
			if (dp[start][end] == -1) {
				if (isPalindrome(s.substr(start, end - start + 1))) {
					dp[start][end] = 1;
				}
				else {
					dp[start][end] = 0;
				}
			}

			if (dp[start][end] == 1) {
				v.push_back(s.substr(start, end - start + 1));

				check(s, end + 1, answer, dp, v);
				v.pop_back();
			}
		}
	}

	vector<vector<string>> partition(string s) {
		vector<vector<int>> dp(s.length(), vector<int>(s.length(), -1));
		vector<vector<string>> answer;
		vector<string> v;

		check(s, 0, answer, dp, v);

		return answer;
	}
};